ang/civicase/ActivityActions--tags.js

```diff
@@ -0,0 +1,216 @@
+      model.tagSets = prepareTagSetsTree(tags);
+
+      model.selectedGenericTags = [];
+      model.toggleGenericTags = toggleGenericTags;
+
+      return model;
+    }
+
+    /**
+     * Toggle the State of Generic tags
+     *
+     * @param {Object} model
+     * @param {String} tagID
+     */
+    function toggleGenericTags (model, tagID) {
+      if (model.selectedGenericTags.indexOf(tagID) === -1) {
+        model.selectedGenericTags.push(tagID);
+      } else {
+        model.selectedGenericTags = _.reject(model.selectedGenericTags, function (tag) {
+          return tag === tagID;
+        });
+      }
+    }
+
+    /**
+     * Opens the modal for addition/removal of tags
+     *
+     * @param {Object} model
+     * @param {String} title
+     * @param {String} saveButtonLabel
+     * @param {String} operation
+     * @param {Array} activities
+     */
+    function openTagsModal (model, title, saveButtonLabel, operation, activities) {
+      dialogService.open('TagsActivityAction', '~/civicase/ActivityActions--tags.html', model, {
+        autoOpen: false,
+        height: 'auto',
+        width: '40%',
+        title: title,
+        buttons: [{
+          text: saveButtonLabel,
+          icons: operation === 'add' ? {primary: 'fa-check'} : false,
+          click: function () {
+            addRemoveTagsConfirmationHandler.call(this, operation, activities, model);
+          }
+        }]
+      });
+    }
+
+    /**
+     * Add/Remove tags confirmation handler
+     *
+     * @param {String} operation
+     * @param {Array} activities
+     * @param {Object} model
+     * @return {Promise}
+     */
+    function addRemoveTagsConfirmationHandler (operation, activities, model) {
+      var tagIds = model.selectedGenericTags;
+
+      _.each(model.tagSets, function (tag) {
+        if (tag.selectedTags) {
+          tagIds = tagIds.concat(JSON.parse('[' + tag.selectedTags + ']'));
+        }
+      });
+
+      var apiCalls = prepareApiCalls(operation, activities, tagIds);
+
+      crmApi(apiCalls)
+        .then(function () {
+          $rootScope.$emit('civicase::activity::updated');
+        });
+
+      $(this).dialog('close');
+    }
+
+    /**
+     * Prepare the API calls for the Add/Remove operation
+     *
+     * @param {String} operation
+     * @param {Array} activities
+     * @param {Array} tagIds
+     * @return {Array}
+     */
+    function prepareApiCalls (operation, activities, tagIds) {
+      var apiCalls = [];
+      var action = operation === 'add' ? 'create' : 'delete';
+
+      _.each(activities, function (activity) {
+        apiCalls.push(['EntityTag', action, {
+          entity_table: 'civicrm_activity',
+          entity_id: activity.id,
+          tag_id: tagIds
+        }]);
+      });
+
+      return apiCalls;
+    }
+    /**
+     * Get the tags for Activities from API end point
+     *
+     * @return {Promise}
+     */
+    function getTags () {
+      return crmApi([['Tag', 'get', {
```

Comment:
Do we need to pass array inside another array. Won't passing a single array element will work for `crmApi`?
Something like 
  ```
  return crmApi(['Tag', 'get', {}]).then.....
```

Reply:
```
crmApi('Tag', 'get', {
        'sequential': 1,
        'used_for': { 'LIKE': '%civicrm_activity%' }
      }).
```
works, thanks for noticing.